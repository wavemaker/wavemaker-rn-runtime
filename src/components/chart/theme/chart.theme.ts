import { assign } from "lodash";
import { VictoryThemeDefinition } from "victory-core";
import ThemeVariables from "@wavemaker/app-rn-runtime/styles/theme.variables";

const themes: {[key:string]: any} = {
  'Terrestrial': {
    colors: ['#1f77b4', '#aec7e8', '#ff7f0e', '#ffbb78', '#2ca02c', '#98df8a', '#d62728', '#ff9896', '#9467bd', '#c5b0d5', '#8c564b', '#c49c94', '#e377c2', '#f7b6d2', '#7f7f7f', '#c7c7c7', '#bcbd22', '#dbdb8d', '#17becf', '#9edae5', '#ccc'],
    tooltip: {
      'backgroundColor': '#de7d28',
      'textColor': '#FFFFFF'
    }
  },
  'Annabelle': {
    colors: ['#393b79', '#5254a3', '#6b6ecf', '#9c9ede', '#637939', '#8ca252', '#b5cf6b', '#cedb9c', '#8c6d31', '#bd9e39', '#e7ba52', '#e7cb94', '#843c39', '#ad494a', '#d6616b', '#e7969c', '#7b4173', '#a55194', '#ce6dbd', '#de9ed6', '#ccc'],
    tooltip: {
      'backgroundColor': '#2e306f',
      'textColor': '#FFFFFF'
    }
  },
  'Azure': {
    colors: ['#3182bd', '#6baed6', '#9ecae1', '#c6dbef', '#e6550d', '#fd8d3c', '#fdae6b', '#fdd0a2', '#31a354', '#74c476', '#a1d99b', '#c7e9c0', '#756bb1', '#9e9ac8', '#bcbddc', '#dadaeb', '#636363', '#969696', '#bdbdbd', '#d9d9d9', '#ccc'],
    tooltip: {
      'backgroundColor': '#3182bd',
      'textColor': '#FFFFFF'
    }
  },
  'Retro': {
    colors: ['#0ca7a1', '#ffa615', '#334957', '#acc5c2', '#988f90', '#8accc9', '#515151', '#f27861', '#36c9fd', '#794668', '#0f709d', '#0d2738', '#44be78', '#4a1839', '#6a393f', '#557d8b', '#6c331c', '#1c1c1c', '#861500', '#09562a', '#ccc'],
    tooltip: {
      'backgroundColor': '#80513a',
      'textColor': '#FFFFFF'
    }
  },
  'Mellow': {
    colors: ['#f0dcbf', '#88c877', '#aeb918', '#2e2c23', '#ddddd2', '#dfe956', '#4c963b', '#5d3801', '#e1eec3', '#cd8472', '#fcfab3', '#9a4635', '#9295ad', '#2e3f12', '#565677', '#557d8b', '#4f4d02', '#0c0c1b', '#833324', '#24120e', '#ccc'],
    tooltip: {
      'backgroundColor': '#7c9e73',
      'textColor': '#FFFFFF'
    }
  },
  'Orient': {
    colors: ['#a80000', '#cc6c3c', '#f0e400', '#000084', '#fccc6c', '#009c6c', '#cc309c', '#78cc00', '#fc84e4', '#48e4fc', '#4878d8', '#186c0c', '#606060', '#a8a8a8', '#000000', '#d7d7d7', '#75a06e', '#190d0b', '#888888', '#694b84', '#ccc'],
    tooltip: {
      'backgroundColor': '#c14242',
      'textColor': '#FFFFFF'
    }
  },
  'GrayScale': {
    colors: ['#141414', '#353535', '#5b5b5b', '#848484', '#a8a8a8', '#c3c3c3', '#e0e0e0', '#c8c8c8', '#a5a5a5', '#878787', '#656565', '#4e4e4e', '#303030', '#1c1c1c', '#4f4f4f', '#3b3b3b', '#757575', '#606060', '#868686', '#c1c1c1', '#ccc'],
    tooltip: {
      'backgroundColor': '#575757',
      'textColor': '#FFFFFF'
    }
  },
  'Flyer': {
    colors: ['#3f454c', '#5a646e', '#848778', '#cededf', '#74c4dd', '#0946ed', '#380bb1', '#000ff0', '#f54a23', '#1db262', '#bca3aa', '#ffa500', '#a86b32', '#63a18c', '#56795e', '#934343', '#b75f5f', '#752d2d', '#4e1111', '#920606', '#ccc'],
    tooltip: {
      'backgroundColor': '#47637c',
      'textColor': '#FFFFFF'
    }
  },
  'Luminosity': {
    colors: ['#FFFFFF', '#e4e4e4', '#00bcd4', '#f0dd2f', '#00aabf', '#018376', '#e91e63', '#39e5d4', '#ff6d6d', '#00ff76', '#ff9800', '#969696', '#ff4200', '#e00000', '#95cbe5', '#5331ff', '#fff4a7', '#e7a800', '#0061e4', '#d5e7ff', '#ccc'],
    tooltip: {
      'backgroundColor': '#47637c',
      'textColor': '#FFFFFF'
    }
  }
}

class ThemeFactory {
  getColorsObj(themeName: string) {
    return themes[themeName].colors;
  }

  getTheme(name: string, styles?: any, customColors?: Array<string>) {
    const colorsToUse = this.getColorsObj(name);
    const [c1, c2, c3, c4, c5, c6, c7, c8, c9, c10] = colorsToUse;
    let colors = [
      c1,
      c2,
      c3,
      c4,
      c5,
      c6,
    ];
    if (customColors && customColors.length) {
      colors = customColors;
    }
    const gridColor = colorsToUse[colorsToUse.length - 1];
    let textColorFromProps;
    if (styles) {
      textColorFromProps = styles.text.color;
    }

    // Typography
    const sansSerif = "'Helvetica Neue', 'Helvetica', sans-serif";
    const letterSpacing = "normal";
    const fontSize = 12;

    // Layout
    const padding = 8;
    const baseProps = {
      width: 350,
      height: 350,
      padding: 50,
    };

    // Labels
    const baseLabelStyles = {
      fontFamily: sansSerif,
      fontSize,
      letterSpacing,
      padding,
      fill: textColorFromProps || ThemeVariables.defaultTextColor,
      stroke: "transparent",
      strokeWidth: 0,
    };

    const centeredLabelStyles = assign({ textAnchor: "middle" }, baseLabelStyles);

    // Strokes
    const strokeDasharray = "10, 5";
    const strokeLinecap = "round";
    const strokeLinejoin = "round";

    const baseChartTheme: VictoryThemeDefinition = {
      area: assign(
        {
          style: {
            data: {
              fill: c10,
            },
            labels: baseLabelStyles,
          },
        },
        baseProps,
      ),
      axis: assign(
        {
          style: {
            axis: {
              fill: "transparent",
              stroke: textColorFromProps || ThemeVariables.chartAxisColor,
              strokeWidth: 2,
              strokeLinecap,
              strokeLinejoin,
            },
            axisLabel: assign({}, centeredLabelStyles, {
              padding: 30,
              stroke: "transparent",
              fontSize: 15
            }),
            grid: {
              fill: "none",
              stroke: gridColor,
              strokeDasharray,
              strokeLinecap,
              strokeLinejoin,
              pointerEvents: "painted",
            },
            ticks: {
              fill: "transparent",
              size: 5,
              stroke: ThemeVariables.chartAxisPointColor,
              strokeWidth: 1,
              strokeLinecap,
              strokeLinejoin,
            },
            tickLabels: assign({}, baseLabelStyles, {
              fill: textColorFromProps || ThemeVariables.defaultTextColor,
            }),
          },
        },
        baseProps,
      ),
      polarDependentAxis: assign({
        style: {
          ticks: {
            fill: "transparent",
            size: 1,
            stroke: "transparent",
          },
        },
      }),
      bar: assign(
        {
          style: {
            data: {
              fill: c1,
              padding,
              strokeWidth: 0,
            },
            labels: baseLabelStyles,
          },
        },
        baseProps,
      ),
      boxplot: assign(
        {
          style: {
            max: { padding, stroke: c9, strokeWidth: 1 },
            maxLabels: assign({}, baseLabelStyles, { padding: 3 }),
            median: { padding, stroke: c9, strokeWidth: 1 },
            medianLabels: assign({}, baseLabelStyles, { padding: 3 }),
            min: { padding, stroke: c9, strokeWidth: 1 },
            minLabels: assign({}, baseLabelStyles, { padding: 3 }),
            q1: { padding, fill: c9 },
            q1Labels: assign({}, baseLabelStyles, { padding: 3 }),
            q3: { padding, fill: c9 },
            q3Labels: assign({}, baseLabelStyles, { padding: 3 }),
          },
          boxWidth: 20,
        },
        baseProps,
      ),
      candlestick: assign(
        {
          style: {
            data: {
              stroke: c9,
            },
            labels: assign({}, baseLabelStyles, { padding: 5 }),
          },
          candleColors: {
            positive: "#ffffff",
            negative: c9,
          },
        },
        baseProps,
      ),
      chart: baseProps,
      errorbar: assign(
        {
          borderWidth: 8,
          style: {
            data: {
              fill: "transparent",
              opacity: 1,
              stroke: c9,
              strokeWidth: 2,
            },
            labels: baseLabelStyles,
          },
        },
        baseProps,
      ),
      group: assign(
        {
          colorScale: colors,
        },
        baseProps,
      ),
      histogram: assign(
        {
          style: {
            data: {
              fill: c9,
              stroke: c10,
              strokeWidth: 2,
            },
            labels: baseLabelStyles,
          },
        },
        baseProps,
      ),
      legend: {
        colorScale: colors,
        gutter: 10,
        orientation: "vertical",
        titleOrientation: "top",
        style: {
          data: {
            type: "circle",
          },
          border: { stroke: ThemeVariables.chartLegendBorder },
          labels: baseLabelStyles,
          title: assign({}, baseLabelStyles, { padding: 5 }),
        },
      },
      line: assign(
        {
          style: {
            data: {
              fill: "transparent",
              opacity: 1,
              stroke: c1,
              strokeWidth: 2,
            },
            labels: baseLabelStyles,
          },
        },
        baseProps,
      ),
      pie: assign(
        {
          colorScale: colors,
          style: {
            data: {
              padding,
              stroke: '#ffffff',
              strokeWidth: 1,
            },
            labels: assign({}, baseLabelStyles, { padding: 20 }),
          },
        },
        baseProps,
      ),
      scatter: assign(
        {
          style: {
            data: {
              fill: c9,
              opacity: 1,
              stroke: "transparent",
              strokeWidth: 0,
            },
            labels: baseLabelStyles,
          },
        },
        baseProps,
      ),
      stack: assign(
        {
          colorScale: colors,
        },
        baseProps,
      ),
      tooltip: {
        style: assign({}, baseLabelStyles, { padding: 0, pointerEvents: "none" }),
        flyoutStyle: {
          stroke: c10,
          strokeWidth: 1,
          fill: "#f0f0f0",
          pointerEvents: "none",
        },
        flyoutPadding: 5,
        cornerRadius: 5,
        pointerLength: 10,
      },
      voronoi: assign(
        {
          style: {
            data: {
              fill: "transparent",
              stroke: "transparent",
              strokeWidth: 0,
            },
            labels: assign({}, baseLabelStyles, {
              padding: 5,
              pointerEvents: "none",
            }),
            flyout: {
              stroke: c10,
              strokeWidth: 1,
              fill: "#f0f0f0",
              pointerEvents: "none",
            },
          },
        },
        baseProps,
      ),
    };
    return baseChartTheme;
  }
}

export default new ThemeFactory();
