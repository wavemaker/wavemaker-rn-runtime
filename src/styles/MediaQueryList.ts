import mediaQuery from "css-mediaquery";
import { Dimensions } from "react-native";
import viewport from "@wavemaker/app-rn-runtime/core/viewport";

export default class MediaQueryList /* extends MediaQueryList */ {

  constructor(private query: string) {}

  public get matches(): boolean {
    const windowDimensions = Dimensions.get("window");
    return mediaQuery.match(this.query, {
      type: "screen",
      orientation: viewport.orientation,
      ...windowDimensions,
      "device-width": viewport.width,
      "device-height": viewport.height,
    } as any);
  }
}