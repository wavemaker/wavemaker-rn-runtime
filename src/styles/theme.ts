import { deepCopy } from '@wavemaker/app-rn-runtime/core/utils';
import { merge } from 'lodash';
import React from 'react';
import { TextStyle, ViewStyle, ImageStyle } from 'react-native';
export const DEFAULT_CLASS = 'DEFAULT_CLASS';
export const DEFAULT_STYLE: NamedStyles<any> = {};

export class Theme {
    private styles: any = {};

    private cache: any = {};

    public static BASE = new Theme();

    private constructor(private parent?:Theme) {

    }

    addStyle<T extends NamedStyles<any>>(name: string, extend: string, style: T) {
        this.styles[name] = deepCopy({}, this.styles[extend], this.styles[name], style);
    }
    
    getStyle(name: string) {
        let style = this.cache[name];
        if (style) {
            return style;
        }
        if (!name) {
            return {};
        }
        if (name.indexOf(' ') > 0) {
            style = merge({}, ...(name.split(' ').map(c => this.getStyle(c))));            
        } else {
            style = merge({}, this.parent && this.parent.getStyle(name), this.styles[name]);
        }
        this.cache[name] = style;
        return style;
    }

    $new(styles = DEFAULT_STYLE) {
        const newTheme = new Theme(this);
        Object.keys(styles).forEach(k => {
            newTheme.addStyle(k, '', styles[k] as any);
        });
        return newTheme;
    }
}
export default Theme.BASE;
export type NamedStyles<T> = { [P in keyof T]: ViewStyle | TextStyle | ImageStyle | NamedStyles<T>};
export type AllStyle = ViewStyle | TextStyle | ImageStyle;

const ThemeContext = React.createContext<Theme>(null as any);

export const ThemeProvider = ThemeContext.Provider;
export const ThemeConsumer = ThemeContext.Consumer;