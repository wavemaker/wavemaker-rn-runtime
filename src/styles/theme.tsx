import { cloneDeep, isNil, forEach, flatten, isArray, isEmpty, isObject, isString, isFunction, get, reverse } from 'lodash';
import React from 'react';
import { camelCase } from 'lodash-es';
import { TextStyle, ViewStyle, ImageStyle, ImageBackground, Dimensions } from 'react-native';
import { deepCopy, isWebPreviewMode } from '@wavemaker/app-rn-runtime/core/utils';
import EventNotifier from '@wavemaker/app-rn-runtime/core/event-notifier';
import ViewPort, {EVENTS as ViewPortEvents} from '@wavemaker/app-rn-runtime/core/viewport';
import MediaQueryList from './MediaQueryList';
import ThemeVariables from './theme.variables';
import { getErrorMessage, getStyleReference, isValidStyleProp } from './style-prop.validator';
export const DEFAULT_CLASS = 'DEFAULT_CLASS';

declare const matchMedia: any, window: any;

if (typeof window !== "undefined") {
    // @ts-ignore: does not properly extend MediaQueryList
    window.matchMedia = (query: string) => new MediaQueryList(query);
}

export const DEVICE_BREAK_POINTS = {
    'MIN_EXTRA_SMALL_DEVICE' : '0px',
    'MAX_EXTRA_SMALL_DEVICE' : '767px',
    'MIN_SMALL_DEVICE' : '768px',
    'MAX_SMALL_DEVICE' : '991px',
    'MIN_MEDIUM_DEVICE' : '992px',
    'MAX_MEDIUM_DEVICE' : '1199px',
    'MIN_LARGE_DEVICE' : '1200px',
    'MAX_LARGE_DEVICE' : '1000000px',
};

export type styleGeneratorFn<T extends NamedStyles<any>> = (
    themeVariables: ThemeVariables,
    addStyle: (name: string, extend: string, style: T) => void) => void

export enum ThemeEvent {
    CHANGE ='change'
};

export class Theme {
    public static BASE = new Theme(null as any, 'default');

    static {
        ViewPort.subscribe(ViewPortEvents.SIZE_CHANGE, () => {
            Theme.BASE.reset();
        });
    }


    private eventNotifer = new EventNotifier();

    private children: Theme[] = [];

    private styles: any = {};

    private cache: any = {};

    private traceEnabled = false;

    private styleGenerators: styleGeneratorFn<any>[] = [];

    private constructor(private parent:Theme, public readonly name: string) {
        if (parent) {
            this.traceEnabled = parent.traceEnabled;
        } else {
            this.traceEnabled = isWebPreviewMode();
        }
    }

    public subscribe(event: ThemeEvent, fn: Function) {
        return this.eventNotifer.subscribe(event, fn);
    }

    public notify(event: ThemeEvent): void {
        this.eventNotifer.notify(event, []);
        this.children.forEach(t => t.notify(event));
    }

    private replaceVariables(val: string) {
        if(isString(val)) { 
            (val.match(/_*var\([^\)]*\)/g) || []).forEach((s) => {
                const variableName = s.substring(4, s.length - 1);
                val = val.replace(s, (ThemeVariables.INSTANCE as any)[variableName]
                    || (ThemeVariables.INSTANCE as any)[variableName.substring(2)]
                    || (ThemeVariables.INSTANCE as any)[camelCase(variableName.substring(2))]);
                val = this.replaceVariables(val);
            });
        }
        return val; 
    }

    clearCache() {
        this.cache = {};
        this.children.forEach((t) => t.clearCache());
    }

    registerStyle<T extends NamedStyles<any>>(fn: styleGeneratorFn<T>) {
        this.styleGenerators.push(fn);
        fn(ThemeVariables.INSTANCE, this.addStyle.bind(this));
    }

    checkStyleProperties(name: string, value : any) {
        if (isObject(value)) {
            Object.keys(value).map((k) => this.checkStyleProperties(k, (value as any)[k]));
        } else if(name && !isValidStyleProp(name, value)) {
            console.log(
                `%cInvalid Style property in ${this.name}: ${getErrorMessage(name, value)}`,
                'background-color: #FF0000;font-weight: bold; color: #fff'
            );
            console.log(`Refer: ${getStyleReference(name)}`);
        }
    }

    private addStyle<T extends NamedStyles<any>>(name: string, extend: string, style: T) {
        this.styles[name] = deepCopy(this.getStyle(extend), this.styles[name], style);
    }

    private addTrace(styleName: string, mergedChildstyle: any, childStyle: any, parentStyle?: any) {
        if (!this.traceEnabled) {
            return;
        }
        let addTrace = !isEmpty(childStyle);
        forEach(mergedChildstyle, (v: any, k: string) => {
            if (v && !isString(v) && !isArray(v) && isObject(v)) {
                addTrace = false;
                this.addTrace(styleName + '.' + k, v, childStyle && childStyle[k], parentStyle && parentStyle[k])
            }
        });
        if (addTrace) {
            mergedChildstyle['__trace'] = [
                {
                    name: styleName,
                    value: childStyle
                },
                ...(parentStyle?.__trace|| [])
            ];
        } else {
            mergedChildstyle['__trace'] = [...(parentStyle?.__trace|| [])];
        }
    }

    private flatten(style: any, prefix = "", result = {} as any) {
        let collect = !isEmpty(style);
        forEach(style, (v: any, k: string) => {
            if (v && !isString(v) && !isArray(v) && isObject(v)) {
                collect = false;
                this.flatten(v, (prefix ?  prefix + '.' : '') + k, result)
            }
        });
        if (collect) {
            result[prefix] = style;
        }
        return result;
    }

    mergeStyle(...styles: any) {
        const style = deepCopy(...styles);
        if (this.traceEnabled) {
            const flattenStyles = this.flatten(style);
            Object.keys(flattenStyles).forEach(k => {
                const s = flattenStyles[k];
                s['__trace'] = flatten(reverse(styles.map((v: any) => {
                    const cs = get(v, k);
                    if (cs && cs.__trace) {
                        return [...cs.__trace];
                    }
                    return [];
                }).filter((t: any) => t.length > 0)));
            });
        }
        return style;
    }

    cleanseStyleProperties(style: any) {
        if (!(style && isObject(style)) || isString(style) || isArray(style)) {
            return;
        }
        style = style as any;
        if (isObject(style) && !isArray(style)) {
            Object.keys(style).forEach(k => {
                (style as any)[k] = this.replaceVariables((style as any)[k]);
            });
        }
        if (!isNil(style['shadowRadius'])) {
            if (style['shadowRadius'] <= 0) {
                style['shadowColor'] = 'transparent';
            } else if (isNil(style['elevation'])) {
                style['elevation'] = 2;
            }
        }
        if (!isNil(style['margin'])) {
            style['marginLeft'] = style['marginLeft'] || style['margin'];
            style['marginRight'] = style['marginRight'] || style['margin'];
            style['marginTop'] = style['marginTop'] || style['margin'];
            style['marginBottom'] = style['marginBottom'] || style['margin'];
            delete style['margin'];
        }
        if (!isNil(style['padding'])) {
            style['paddingLeft'] = style['paddingLeft'] || style['padding'];
            style['paddingRight'] = style['paddingRight'] || style['padding'];
            style['paddingTop'] = style['paddingTop'] || style['padding'];
            style['paddingBottom'] = style['paddingBottom'] || style['padding'];
            delete style['padding'];
        }
        let screenWidth = Dimensions.get('window').width;
        let screenHeight = Dimensions.get('window').height;
        Object.keys(style).forEach((k, i) => {
            let stylePropertyValue = style[k]
            if(typeof stylePropertyValue === 'string' && stylePropertyValue.endsWith('vw')) {
                stylePropertyValue = stylePropertyValue.replace('vw','')
                style[k] = Number(stylePropertyValue)/100 * screenWidth
            }
            if((typeof stylePropertyValue === 'string' && stylePropertyValue.endsWith('vh'))) {
                stylePropertyValue = stylePropertyValue.replace('vh','')
                style[k] = Number(stylePropertyValue)/100 * screenHeight
            }
        })
        Object.keys(style).forEach((k, i) => this.cleanseStyleProperties(style[k]));
    }

    getStyle(name: string) {
        let style = this.cache[name];
        if (style) {
            return style;
        }
        if (!name) {
            return {};
        }
        if (name.indexOf(' ') > 0) {
            style = this.mergeStyle(...(name.split(' ').map(c => this.getStyle(c))));
        } else {
            const parentStyle = this.parent && this.parent.getStyle(name);
            const mediaQuery = (this.styles[name] || {})['@media'];
            let clonedStyle = {};
            if (!mediaQuery || matchMedia(mediaQuery).matches) {
                clonedStyle = cloneDeep(this.styles[name]);
                this.cleanseStyleProperties(clonedStyle);
            }
            if (this !== Theme.BASE && isWebPreviewMode()) {
                this.checkStyleProperties('', clonedStyle);
            }
            style = deepCopy(parentStyle, clonedStyle);
            this.addTrace(`@${this.name}:${name}`, style, clonedStyle, parentStyle);
        }
        this.cache[name] = style;
        return style;
    }

    $new(name = "", styles = {} as NamedStyles<any>) {
        const newTheme = new Theme(this, name);
        newTheme.reset(styles);
        this.children.push(newTheme);
        return newTheme;
    }

    destroy() {
        const i = this.parent.children.indexOf(this);
        if (i >= 0) {
            this.parent.children.splice(i, 1);
        }
    }
    
    getTextStyle(s: any) {
        if (!s) {
            return {};
        }
        return {
            color: s.color,
            fontFamily: s.fontFamily,
            fontSize: s.fontSize,
            fontStyle: s.fontStyle,
            fontWeight: s.fontWeight,
            includeFontPadding: s.includeFontPadding,
            fontVariant: s.fontVariant,
            letterSpacing: s.letterSpacing,
            lineHeight: s.lineHeight,
            textAlign: s.textAlign,
            textAlignVertical: s.textAlignVertical,
            textDecorationColor: s.textDecorationColor,
            textDecorationStyle: s.textDecorationStyle,
            textShadowColor: s.textShadowColor,
            textShadowOffset: s.textShadowOffset,
            textShadowRadius: s.textShadowRadius,
            textTransform: s.textTransform,
            verticalAlign: s.verticalAlign,
            writingDirection: s.writingDirection,
            userSelect: s.userSelect,
        } as TextStyle;
    }

    reset(styles?: NamedStyles<any>) {
        this.styles = {};
        this.clearCache();
        if (styles) {
            this.registerStyle((themeVariables, addStyle) => {
                Object.keys(styles).forEach(k => {
                    addStyle(k, '', styles[k] as any);
                });
            });
        } else {
            this.styleGenerators.forEach(fn => 
                fn(ThemeVariables.INSTANCE, this.addStyle.bind(this)));
        }
        this.notify(ThemeEvent.CHANGE);
    }
}
export default Theme.BASE;
export type NamedStyles<T> = { [P in keyof T]: AllStyle | NamedStyles<T>};
export type BackgroundImageStyle = {
    backgroundImage: string,
    backgroundPosition: string,
    backgroundRepeat: string,
    backgroundSize: string | number
};

export type AllStyle = (ViewStyle & TextStyle & ImageStyle & {userSelect?: 'none'| 'text'} & {rippleColor?: string});

const ThemeContext = React.createContext<Theme>(null as any);

export const ThemeProvider = ThemeContext.Provider;
export const ThemeConsumer = ThemeContext.Consumer;
/**
 * Common styles
 */
 Theme.BASE.registerStyle((themeVariables, addStyle) => {
    const addColStyles = (device: string, minWidth: string) => {
        for(let i = 1; i <= 12; i++) {
            addStyle(`col-${device}-${i}`, '', {
                "@media": `(min-width: ${minWidth})`,
                root: {
                    width: (100 * i / 12) + '%'
                }
            } as any)
        }
    };
    addColStyles('xs', DEVICE_BREAK_POINTS.MIN_EXTRA_SMALL_DEVICE);
    addColStyles('sm',  DEVICE_BREAK_POINTS.MIN_SMALL_DEVICE);
    addColStyles('md',  DEVICE_BREAK_POINTS.MIN_MEDIUM_DEVICE);
    addColStyles('lg',  DEVICE_BREAK_POINTS.MIN_LARGE_DEVICE);

    addStyle('d-none', '', {
        root: {
            display: 'none'
        }
    } as any);
    addStyle('d-flex', '', {
        root: {
            display: 'flex'
        }
    } as any);

    const addDisplayStyles = (device: string, minWidth: string, maxWidth: string) => {
        addStyle(`d-${device}-none`, '', {
            "@media": `(min-width: ${minWidth}) and (max-width: ${maxWidth})`,
            root: {
                display: 'none'
            }
        } as any);
        addStyle(`d-${device}-flex`, '', {
            "@media": `(min-width: ${minWidth}) and (max-width: ${maxWidth})`,
            root: {
                display: 'flex'
            }
        } as any);
    };
    addDisplayStyles('all', 
        DEVICE_BREAK_POINTS.MIN_EXTRA_SMALL_DEVICE,
        DEVICE_BREAK_POINTS.MAX_LARGE_DEVICE);
    addDisplayStyles('xs', 
        DEVICE_BREAK_POINTS.MIN_EXTRA_SMALL_DEVICE,
        DEVICE_BREAK_POINTS.MAX_EXTRA_SMALL_DEVICE);
    addDisplayStyles('sm',   
        DEVICE_BREAK_POINTS.MIN_SMALL_DEVICE,
        DEVICE_BREAK_POINTS.MAX_SMALL_DEVICE);
    addDisplayStyles('md',   
        DEVICE_BREAK_POINTS.MIN_MEDIUM_DEVICE,
        DEVICE_BREAK_POINTS.MAX_MEDIUM_DEVICE);
    addDisplayStyles('lg',   
        DEVICE_BREAK_POINTS.MIN_LARGE_DEVICE,
        DEVICE_BREAK_POINTS.MAX_LARGE_DEVICE);

    const addElevationClasses = () => {
        for(let i = 1; i <= 10; i++) {
            addStyle(`elevate${i}`, '', {
                root : {
                    shadowColor: "#000000",
                    shadowOffset: {
                        width: i,
                        height: i,
                    },
                    shadowOpacity: 0.27,
                    shadowRadius: i,
                    elevation: i,
                    zIndex: 1
                }
            });
        }
    };
    addElevationClasses();
    addStyle('hidden', '', {
        root: {
            width: 0,
            height: 0,
            transform: [{ scale: 0 }]
        }
    });
    addStyle('bg-danger', '', { root: { backgroundColor: themeVariables.dangerColor }});
    addStyle('bg-info', '', { root: { backgroundColor: themeVariables.infoColor }});
    addStyle('bg-primary', '', { root: { backgroundColor: themeVariables.primaryColor }});
    addStyle('bg-success', '', { root: { backgroundColor: themeVariables.successColor }});
    addStyle('bg-warning', '', { root: { backgroundColor: themeVariables.warningColor }});

    addStyle('border-danger', '', { root: { borderColor: themeVariables.dangerColor }});
    addStyle('border-info', '', { root: { borderColor: themeVariables.infoColor }});
    addStyle('border-primary', '', { root: { borderColor: themeVariables.primaryColor }});
    addStyle('border-success', '', { root: { borderColor: themeVariables.successColor }});
    addStyle('border-warning', '', { root: { borderColor: themeVariables.warningColor }});

    addStyle('hide-context-menu', '', { text: { userSelect: 'none' }});
});

