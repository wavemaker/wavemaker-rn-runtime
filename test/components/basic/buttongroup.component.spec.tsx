import React from 'react';
import { render } from '@testing-library/react-native';
import WmButtongroupProps from '@wavemaker/app-rn-runtime/components/basic/buttongroup/buttongroup.props';
import WmButtongroup from '@wavemaker/app-rn-runtime/components/basic/buttongroup/buttongroup.component';
import { ThemeProvider } from '@wavemaker/app-rn-runtime/styles/theme';
import WmButton from '@wavemaker/app-rn-runtime/components/basic/button/button.component';
import { Text } from 'react-native';

// Mock theme
const mockTheme = {
  buttonGrpBorderColor: '#000000',
  buttonGrpBgColor: '#ffffff',
};

const renderWithTheme = (component: React.ReactElement) => {
  return render(<ThemeProvider value={mockTheme}>{component}</ThemeProvider>);
};

describe('WmButtongroup', () => {
  it('renders correctly with default props', () => {
    const props = {
      vertical: false,
      name: 'buttongroup1',
    };
    const tree = render(<WmButtongroup {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders with `vertical` prop set to true', () => {
    const props = {
      vertical: true,
    };
    const tree = render(<WmButtongroup {...props} />);
    const buttonGroupRoot = tree.root;

    expect(buttonGroupRoot.props.style).toContainEqual({
      flexDirection: 'column',
    });
  });

  it('renders skeleton when `showskeleton` is true', () => {
    const props = {
      vertical: false,
      showskeleton: true,
      skeletonwidth: 90,
      skeletonheight: 40,
      children: <Text>Child Element</Text>,
    };
    const tree = render(<WmButtongroup {...props} />);
    expect(tree.queryByText('Child Element')).toBeNull();
    expect(tree).toMatchSnapshot();
    expect(tree.root.props.style.height).toBe(props.skeletonheight);
    expect(tree.root.props.style.width).toBe(props.skeletonwidth);
  });

  it('should render skeleton with root width and height when skeleton width and height are not provided', () => {
    const width = 50;
    const height = 70;
    const props = {
      showskeleton: true,
      styles: {
        root: {
          width: width,
          height: height,
        },
      },
    };
    const tree = render(<WmButtongroup {...props} />);
    expect(tree).toMatchSnapshot();
    expect(tree.root.props.style.height).toBe(height);
    expect(tree.root.props.style.width).toBe(width);
  });

  it('should have width and height to be 0 when show is false', () => {
    const props = {
      vertical: false,
      show: false,
    };
    const tree = render(<WmButtongroup {...props} />);
    const buttonGroupRoot = tree.root;

    expect(buttonGroupRoot.props.style[0].width).toBe(0);
    expect(buttonGroupRoot.props.style[0].height).toBe(0);
  });

  it('renders with given width', () => {
    const width = 70;
    const props = {
      vertical: false,
      styles: {
        root: {
          width: width,
        },
      },
    };

    const tree = render(<WmButtongroup {...props} />);
    const buttonGroupRoot = tree.root;
    expect(buttonGroupRoot.props.style[0].width).toBe(width);
  });

  it('renders children correctly', () => {
    const props = {
      vertical: false,
    };
    const { getByText } = render(
      <WmButtongroup {...props}>
        <WmButton caption="Button1" />
        <WmButton caption="Button2" />
      </WmButtongroup>
    );

    expect(getByText('Button1')).toBeTruthy();
    expect(getByText('Button2')).toBeTruthy();
  });

  it('should render custom styles with properly', () => {
    const bgColor = '#234555';
    const props = {
      styles: {
        root: {
          backgroundColor: bgColor,
        },
      },
    };

    const tree = render(<WmButtongroup {...props} />);
    const buttonGroupRoot = tree.root;
    expect(buttonGroupRoot.props.style[0].backgroundColor).toBe(bgColor);
  });

  xit('applies styles correctly based on props and themes', () => {
    const props = {
      vertical: false,
      children: <></>,
    };

    const tree = render(<WmButtongroup {...props} />);
    const buttonGroupRoot = tree.root;

    expect(buttonGroupRoot.props.style).toContainEqual({
      borderColor: mockTheme.buttonGrpBorderColor,
      backgroundColor: mockTheme.buttonGrpBgColor,
    });
  });
});
