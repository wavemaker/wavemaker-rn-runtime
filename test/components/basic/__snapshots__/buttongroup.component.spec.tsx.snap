// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WmButtongroup renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#ffffff",
        "borderColor": "#cccccc",
        "borderRadius": 100,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "overflow": "hidden",
      },
      {
        "flexDirection": "row",
      },
    ]
  }
/>
`;

exports[`WmButtongroup renders skeleton when \`showskeleton\` is true 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "backgroundColor": "#eeeeee",
      "borderColor": "#cccccc",
      "borderRadius": 100,
      "height": 40,
      "overflow": "hidden",
      "width": 90,
    }
  }
/>
`;

exports[`WmButtongroup should render skeleton with root width and height when skeleton width and height are not provided 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "backgroundColor": "#eeeeee",
      "borderColor": "#cccccc",
      "borderRadius": 100,
      "height": 70,
      "overflow": "hidden",
      "width": 50,
    }
  }
/>
`;
